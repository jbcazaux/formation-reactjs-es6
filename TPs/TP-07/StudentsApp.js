import React, { useEffect, useState } from 'react'
import axios from 'axios'
import StudentFilter from './StudentFilter'
import StudentsTable from './StudentsTable'
import StudentDetails from './StudentDetails'

const filteredStudents = (students, filter) =>
  students.filter(s => s.firstname.includes(filter) || s.lastname.includes(filter))

const StudentsApp = () => {
  const [students, setStudents] = useState([])
  const [filter, setFilter] = useState('')
  const [selectedStudent, setSelectedStudent] = useState(null)

  useEffect(() => {
    axios.get('students.json').then(({ data: students }) => setStudents(students))
  }, [])

  const handleFilterChange = f => {
    setFilter(f)
  }

  const handleSelectStudent = s => {
    setSelectedStudent(s)
  }

  return (
    <>
      <StudentFilter onChange={handleFilterChange} />
      <StudentsTable students={filteredStudents(students, filter)} selectStudent={handleSelectStudent} />
      <StudentDetails student={selectedStudent} />
    </>
  )
}

export default StudentsApp
