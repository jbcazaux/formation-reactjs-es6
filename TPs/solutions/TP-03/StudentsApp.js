import { useEffect, useState } from 'react'
import axios from 'axios'
import StudentFilter from './StudentFilter'
import StudentsTable from './StudentsTable'
import StudentDetails from './StudentDetails'

const filterStudents = (students, filter) => {
  return students.filter(s => s.firstname.includes(filter) || s.lastname.includes(filter))
}

const StudentsApp = () => {
  const [students, setStudents] = useState([])
  const [filter, setFilter] = useState('')
  const [student, setStudent] = useState(null)

  useEffect(() => {
    axios.get('students.json').then(({ data }) => {
      setStudents(data)
    })
  }, [])

  const handleFilterChange = newFilter => {
    setFilter(newFilter)
  }

  const handleSelectStudent = newSelectedStudent => {
    setStudent(newSelectedStudent)
  }

  return (
    <>
      <StudentFilter onChange={handleFilterChange} />
      <StudentsTable students={filterStudents(students, filter)} onSelectStudent={handleSelectStudent} />
      <StudentDetails student={student} />
    </>
  )
}

export default StudentsApp
