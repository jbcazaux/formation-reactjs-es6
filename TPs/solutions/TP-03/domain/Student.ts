class Student {
  constructor(
    readonly id: number,
    readonly lastname: string,
    readonly firstname: string,
    readonly grades: ReadonlyArray<number>
  ) {}
}

export default Student
