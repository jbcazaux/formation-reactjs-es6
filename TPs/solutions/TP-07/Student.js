import PropTypes from 'prop-types'

class Student {
  constructor(id, lastname, firstname, grades) {
    this.id = id
    this.lastname = lastname
    this.firstname = firstname
    this.grades = grades
  }
}

Student.propTypes = PropTypes.shape({
  firstname: PropTypes.string.isRequired,
  grades: PropTypes.arrayOf(PropTypes.number).isRequired,
  id: PropTypes.number.isRequired,
  lastname: PropTypes.string.isRequired,
})

export default Student
