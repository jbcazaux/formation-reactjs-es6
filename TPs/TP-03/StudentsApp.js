import React, { useEffect, useState } from 'react'
import axios from 'axios'
import StudentFilter from './StudentFilter'
import StudentsTable from './StudentsTable'
import StudentDetails from './StudentDetails'
import Student from './Student'

const filterStudents = (students, filter) => {
  // TODO: retourner une liste filrée des étudiants, plutot que la liste entière
  return students
}

const StudentsApp = () => {
  // TODO créer un state pour stocker la liste des étudiants
  // TODO créer un state pour stocker le filtre (c'est une chaine de caractères)
  // TODO créer un state pour stocker l'étudiant actuellement sélectionné (null par défaut)

  useEffect(() => {
    axios.get('students.json').then(({ data }) => {
      // TODO: mettre le résultat de la requete dans le state 'students'
    })
  }, [])

  const handleFilterChange = newFilter => {
    // TODO stocker la nouvelle valeur du filtre
  }

  const handleSelectStudent = newSelectedStudent => {
    // TODO stocker le nouvel étudiant sélectionné
  }

  return (
    <>
      TODO: Décommenter les composants enfants au fur et à mesure
      <StudentFilter onChange={handleFilterChange} />
      <StudentsTable students={filterStudents(students, filter)} onSelectStudent={handleSelectStudent} />
      <StudentDetails student={selectedStudent} />
    </>
  )
}

export default StudentsApp
