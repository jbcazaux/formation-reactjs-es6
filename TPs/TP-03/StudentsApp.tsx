import { useEffect, useState } from 'react'
import axios from 'axios'
import StudentFilter from './StudentFilter'
import StudentsTable from './StudentsTable'
import StudentDetails from './StudentDetails'
import Student from './domain/Student'

const filterStudents = (students: ReadonlyArray<Student>, filter: string): ReadonlyArray<Student> => {
  // TODO: retourner une liste filrée des étudiants, plutot que la liste entière
  return students
}

const StudentsApp = () => {
  // TODO créer un state 'students' pour stocker la liste des étudiants
  // TODO créer un state 'filter' pour stocker le filtre (c'est une chaine de caractères)
  // TODO créer un state 'selectedStudent' pour stocker l'étudiant actuellement sélectionné (null par défaut)

  useEffect(() => {
    axios
      .get<ReadonlyArray<Student>>('students.json')
      .then(({ data }) => data.map(d => new Student(d.id, d.lastname, d.firstname, d.grades)))
      .then(fetchedStudents => {
        // TODO: mettre le résultat de la requete dans le state 'students'
      })
  }, [])

  const handleFilterChange = (newFilter: string) => {
    // TODO stocker la nouvelle valeur du filtre dans son state
  }

  const handleSelectStudent = (newSelectedStudent: Student) => {
    // TODO stocker le nouvel étudiant sélectionné dans son state
  }

  return (
    <>
      TODO: Décommenter les composants enfants au fur et à mesure : StudentsTable, StudentFilter puis StudentDetails
      {/*<StudentFilter onChange={handleFilterChange} />*/}
      {/*<StudentsTable students={students} onSelectStudent={handleSelectStudent} />*/}
      {/*<StudentDetails student={selectedStudent} />*/}
    </>
  )
}

export default StudentsApp
