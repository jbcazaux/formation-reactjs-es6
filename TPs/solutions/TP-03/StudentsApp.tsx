import { useEffect, useState } from 'react'
import axios from 'axios'
import StudentFilter from './StudentFilter'
import StudentsTable from './StudentsTable'
import StudentDetails from './StudentDetails'
import Student from './domain/Student'

const filterStudents = (students: ReadonlyArray<Student>, filter: string): ReadonlyArray<Student> =>
  students.filter(s => s.firstname.includes(filter) || s.lastname.includes(filter))

const StudentsApp = () => {
  const [students, setStudents] = useState<ReadonlyArray<Student>>([])
  const [filter, setFilter] = useState<string>('')
  const [student, setStudent] = useState<Student | null>(null)

  useEffect(() => {
    axios
      .get<ReadonlyArray<Student>>('students.json')
      .then(({ data }) => data.map(d => new Student(d.id, d.lastname, d.firstname, d.grades)))
      .then(setStudents)
  }, [])

  const handleFilterChange = (newFilter: string) => {
    setFilter(newFilter)
  }

  const handleSelectStudent = (newSelectedStudent: Student) => {
    setStudent(newSelectedStudent)
  }

  return (
    <>
      <StudentFilter onChange={handleFilterChange} />
      <StudentsTable students={filterStudents(students, filter)} onSelectStudent={handleSelectStudent} />
      <StudentDetails student={student} />
    </>
  )
}

export default StudentsApp
