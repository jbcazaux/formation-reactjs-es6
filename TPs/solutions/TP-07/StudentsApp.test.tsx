import StudentsApp, { filterStudents } from './StudentsApp'
import { render, screen } from '@testing-library/react'
import axios from 'axios'
import Student from './domain/Student'

describe('StudentsApp', () => {
  beforeEach(() => {
    axios.get = jest.fn().mockImplementationOnce(() => Promise.resolve({ data: [new Student(1, 'last', 'first', [])] }))
  })

  it('renders studentsApp and its sub components', async () => {
    // Given
    const { container } = render(<StudentsApp />)
    await screen.findByTestId('student-row')

    // Then
    expect(container).toMatchSnapshot()
  })

  it('filters students', () => {
    //Given
    const students = [new Student(1, 'last1', 'first1', []), new Student(2, 'last2', 'first2', [])]

    //when
    const result = filterStudents(students, '2')

    //then
    expect(result).toHaveLength(1)
    expect(result[0]).toEqual(students[1])
  })
})
