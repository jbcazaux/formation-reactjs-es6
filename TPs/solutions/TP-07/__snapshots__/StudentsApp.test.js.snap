// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StudentsApp renders studentsApp and its sub components 1`] = `
<div>
  <form>
    <div
      class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary form-control css-q0jhri-MuiInputBase-root-MuiInput-root"
      style="margin-left: 5px;"
    >
      <input
        class="MuiInputBase-input MuiInput-input css-1x51dt5-MuiInputBase-input-MuiInput-input"
        id="filterName"
        placeholder="Tapez un nom..."
        type="text"
        value=""
      />
    </div>
  </form>
  <table
    class="MuiTable-root css-rqglhn-MuiTable-root"
  >
    <thead
      class="MuiTableHead-root css-15wwp11-MuiTableHead-root"
    >
      <tr
        class="MuiTableRow-root MuiTableRow-head css-1q1u3t4-MuiTableRow-root"
      >
        <th
          class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1ygcj2i-MuiTableCell-root"
          scope="col"
        >
          #
        </th>
        <th
          class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1ygcj2i-MuiTableCell-root"
          scope="col"
        >
          Firstname
        </th>
        <th
          class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1ygcj2i-MuiTableCell-root"
          scope="col"
        >
          Lastname
        </th>
      </tr>
    </thead>
    <tbody
      class="MuiTableBody-root css-apqrd9-MuiTableBody-root"
    >
      <tr
        class="MuiTableRow-root css-1q1u3t4-MuiTableRow-root"
        data-testid="student-row"
      >
        <td
          class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
        >
          1
        </td>
        <td
          class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
        >
          first
        </td>
        <td
          class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
        >
          last
        </td>
      </tr>
    </tbody>
  </table>
  <label
    class="MuiFormLabel-root MuiFormLabel-colorPrimary css-u4tvz2-MuiFormLabel-root"
  >
    <div
      style="margin: 20px; background-color: red;"
    >
      Aucun étudiant sélectionné !
    </div>
  </label>
</div>
`;
